src/shims/posix/linux/sync.rs

```diff
@@ -38,23 +40,37 @@ pub fn futex<'tcx>(
     let futex_private = this.eval_libc_i32("FUTEX_PRIVATE_FLAG")?;
     let futex_wait = this.eval_libc_i32("FUTEX_WAIT")?;
     let futex_wake = this.eval_libc_i32("FUTEX_WAKE")?;
+    let futex_realtime = this.eval_libc_i32("FUTEX_CLOCK_REALTIME")?;
 
     // FUTEX_PRIVATE enables an optimization that stops it from working across processes.
     // Miri doesn't support that anyway, so we ignore that flag.
     match op & !futex_private {
         // FUTEX_WAIT: (int *addr, int op = FUTEX_WAIT, int val, const timespec *timeout)
         // Blocks the thread if *addr still equals val. Wakes up when FUTEX_WAKE is called on the same address,
         // or *timeout expires. `timeout == null` for an infinite timeout.
-        op if op == futex_wait => {
+        op if op & !futex_realtime == futex_wait => {
             if args.len() < 5 {
                 throw_ub_format!("incorrect number of arguments for FUTEX_WAIT syscall: got {}, expected at least 5", args.len());
             }
-            let timeout = this.read_scalar(args[4])?.check_init()?;
-            if !this.is_null(timeout)? {
-                // FIXME: Implement timeouts. The condvar waiting code is probably a good example to start with.
-                // Note that a triggered timeout should have this syscall return with -1 and errno set to ETIMEOUT.
-                throw_ub_format!("miri does not support timeouts for futex operations");
-            }
+            let timeout = args[4];
+            let timeout_time = if this.is_null(this.read_scalar(timeout)?.check_init()?)? {
+                None
+            } else {
+                let duration = match this.read_timespec(timeout)? {
```

Comment:
Using timeouts should first check that isolation is disabled, as depending on the clock will break determinism.